fix: close broken tcp connections

Asphaltt opened the pull request:
Prevent TCP connection leaks that error occurred to kernel space's socket. Previously, when errors occurred on TCP sockets, connections would remain in TCP_CLOSE state indefinitely without being properly freed.

The leaked connections caused a cascade of kernel-level issues:
- TCP sockets remain in TCP_CLOSE state indefinitely
- Their sk_receive_queues retain FINACK skb packets
- These skbs hold pages allocated from page_pool
- page_pool_release_retry() stalls because pages cannot be freed while still referenced by the TCP stack

The fix ensures that TCP connections are properly closed when socket errors occur, preventing resource leaks at both the application and kernel levels.

Asphaltt committed:
fix: close broken tcp connections

Prevent TCP connection leaks that error occurred to kernel space's socket.
Previously, when errors occurred on TCP sockets, connections would
remain in TCP_CLOSE state indefinitely without being properly freed.

The leaked connections caused a cascade of kernel-level issues:
- TCP sockets remain in TCP_CLOSE state indefinitely
- Their sk_receive_queues retain FINACK skb packets
- These skbs hold pages allocated from page_pool
- page_pool_release_retry() stalls because pages cannot be freed
  while still referenced by the TCP stack

The fix ensures that TCP connections are properly closed when socket
errors occur, preventing resource leaks at both the application and
kernel levels.

Signed-off-by: Leon Hwang <leon.hwang@linux.dev>

dnwe reviewed (commented):
Thanks for reporting and investigating this issue, as well as raising a PR!

A few thoughts:

Asphaltt commented on client.go:
```diff
@@ -684,6 +684,38 @@ func (client *client) randomizeSeedBrokers(addrs []string) {
+	client.seedBrokers = client.checkSeedBrokersHealth(client.seedBrokers)
+	client.deadSeeds = client.checkSeedBrokersHealth(client.deadSeeds)
+}
+
```
Thanks for your explanation.

> If we delete from seedBrokers (bootstraps) what re-populates them?

In that case, should the seedBrokers be re-opened instead of being removed when we detect `SO_ERROR`?

> why this new health check is bolted on as a separate pass

I'm not entirely sure where the most appropriate place is to perform this health check, as I'm not deeply familiar with all parts of this library. I initially attempted to implement it inside `updateBrokers()`, but the issue still occurred.

With this PR, broken TCP connections would at least be detected and closed every 10 minutes.

> During LeastLoadedBroker/Leader/RefreshMetadata, when a broker returns an error, we already close and deregister it so we could put a check for socket errors there I think?

I'm not certain that covers all cases. For brokers that rarely get used and don't fetch data for a long time, their TCP connections may remain broken for an extended period without triggering any of those paths.

puellanivis commented on internal/sockopt/sockopt_windows.go:
```diff
@@ -0,0 +1,9 @@
+//go:build windows
```
`plan9` also doesn’t have `SO_ERROR`, or `GetSockoptInt`. (Though, in this case, I don’t think we support it, as it already doesn’t build because `go-metrics` tries to use a non-existent `syslog.Writer`.)

Also, neither `js/wasm` nor `wasip1/wasm` work. These ones at least seem to compile prior to this change.

You may want:
```suggestion
//go:build windows || wasm
```

Though, this will require renaming the file, as the `_windows.go` suffix means this build tag for `windows` is currently redundant, and unnecessary.

puellanivis commented on broker.go:
```diff
@@ -156,6 +159,39 @@ func NewBroker(addr string) *Broker {
+		return syscall.Errno(sockErr)
+	}
+	return nil
+}
```
As mentioned, using `SyscallConn` and `RawConn.Control` remove the need to `dup` the underlying file descriptor: 
```golang
func (b *Broker) getSockError() error {
	b.lock.Lock()
	defer b.lock.Unlock()

	if conn, _ := b.conn.(*net.TCPConn); conn != nil {
		rawConn, err := conn.SyscallConn()
		if err != nil {
			return err
		}

		var sockErr int
		var opErr error

		err = rawConn.Control(func(fd uintptr) {
				sockErr, opErr = sockopt.GetSockoptInt(fd, sockopt.SockoptError)
		})
		if err != nil {
			return fmt.Errorf("failed to get socket error: %w", err)
		}

		if opErr != nil {
			return fmt.Errorf("failed to get socket error: %w", err)
		}

		if sockErr != 0 {
			return syscall.Errno(sockErr)
		}
	}

	return nil
}
```

Asphaltt replied:
LGTM.

I'll update it like this.

Thanks!

puellanivis commented on broker.go:
```diff
@@ -156,6 +157,17 @@ func NewBroker(addr string) *Broker {
+		return nil
+	}
+
+	return getTCPConnSockError(b.connTCP)
```
I continue to have concerns about using a different variable to store the same value that would already be available via `b.conn.(*net.TCPConn)`.

There is no need to coördinate handling both the `conn` and `connTCP` variables to ensure that they remain consistent with each other. In context of the [Swiss-Cheese Model](https://en.wikipedia.org/wiki/Swiss_cheese_model) of accident causation, forcing developers to manually ensure the consistency of two variables when it can be driven by just the one variable would open up a significant hole in one of the layers of swiss cheese, making accidents during refactoring more likely.

```suggestion
	connTCP, _ := b.conn.(*net.TCPConn)
	if connTCP == nil {
		return nil
	}

	return getTCPConnSockError(connTCP)
```

puellanivis reviewed (commented):
Looks good just one small possible improvement. (The code, as is, is fine though.)

Asphaltt commented on broker.go:
```diff
@@ -156,6 +156,24 @@ func NewBroker(addr string) *Broker {
+	}
+	if c, ok := conn.(*net.TCPConn); ok {
+		return getTCPConnSockError(c)
+	}
```
Thanks for the suggestion! I appreciate the insight about typed nil pointers.

However, I prefer to keep the current implementation for the following reasons:

1. **Explicit intent**: The `if b.conn == nil` check at the beginning makes it immediately clear that we're guarding against nil connections. While it's true that type assertions will fail on untyped nil, this explicit check documents the precondition and makes the code easier to understand at first glance.

2. **Readability**: Using `ok` in type assertions is a common Go idiom that signals "we're checking if this conversion is safe." Testing `c != nil` after discarding `ok` is less conventional and may confuse readers who expect to see the standard `if x, ok := ...` pattern.

3. **Future-proofing**: If `NetConn()` or the connection wrapping logic changes in the future, the explicit checks make it easier to understand the control flow and add debugging.

While your suggestion is technically correct and slightly more concise, I believe the current version optimizes for clarity and maintainability, which is especially important for connection management code where bugs can be subtle and hard to debug.

What do you think? If there's a strong preference in the project for the more compact style, I'm happy to adjust.

puellanivis replied:
Yes, I am already well aware of the `ok` pattern, and the nuances behind it, which is specifically why I have been bringing up this point.

Testing that the type-assertion succeeded is not always sufficient—as specifically in this case, a type-nil pointer being in the interface will result in a panic on the receiver method calls. So, we should also be testing for validity of the typed pointer `c`: `c != nil`. So, at the very minimum, we should have `ok && c != nil`, which is a test for success and validity.

I don’t see how my code would be any less future proof than yours. First, the go1 compatibility guarantee is in effect for `crypto/tls`, so the `NetConn` behavior shall not change radically enough to force either into breaking.

Asphaltt commented:
PING

---

Over time, `go.mod` and `go.sum` may conflict with the files on the main branch. When that happens, I will need to resolve those conflicts.

Asphaltt commented:
Hi @dnwe, do you have any remaining concerns that need further discussion?

If not, could you please approve this PR?

dnwe committed:
fix: keep bootstrap brokers and unblock async shutdown

Skip socket health checks while a broker is still opening so metadata refresh does not block behind the broker mutex.

Keep seed and dead-seed brokers in their bootstrap pools after socket errors so they can be reopened on a later refresh.

Close broker sockets before waiting for the response reader so health-check cleanup and SASL v1 auth failures cannot hang while async responses are still in flight.

Signed-off-by: Dominic Evans <dominic.evans@uk.ibm.com>

dnwe committed:
fix: skip broker health tests without socket probing

Signed-off-by: Dominic Evans <dominic.evans@uk.ibm.com>

dnwe commented:
@puellanivis can you take a look at the reworked version to see if you agree with my updates? They’re along the lines of what I wrote in https://github.com/IBM/sarama/pull/3384#discussion_r2547361998

puellanivis commented on client.go:
```diff
@@ -684,6 +684,34 @@ func (client *client) randomizeSeedBrokers(addrs []string) {
+		}
+	}
+
+	return brokers
```
I’m not sure what the value is in returning the same slice that was passed in as an argument? I mean, it doesn’t look like there’s any chance that the slice has been modified at all?

One usually only wants to return a value that was passed in as an argument if there’s a chance that it could have reallocated. (like `append` and `slices.Grow`), while otherwise, we just perform the operation on the argument, and don’t return anything at all `slices.Sort`)

puellanivis reviewed (commented):
Seems mostly fine? Just the one thing.

dnwe committed:
fix: remove return from checkSeedBrokersHealth

Signed-off-by: Dominic Evans <dominic.evans@uk.ibm.com>
